src/NSL/FermionMatrix/Impl/hubbardExp.hpp

```diff
@@ -71,29 +71,29 @@ class HubbardExp : public FermionMatrix<Type,LatticeType> {
         phi_ = phi;
 
         // determine the signs for exp(+/- kappa) exp(+/- i phi)
-        int sgn;
+        int sgn = +1;
         if (this->species == NSL::Hubbard::Species::Particle) {
             sgn = +1;
             this->sgn_kappa_ = +1;
-        } else {
+        } else { // species == NSL::Hubbard::Species::Hole
             sgn = -1;
 
             // On bipartite lattices exp(-kappa) -> exp(kappa)
             // else we need a - sign for holes
-            if(!this->Lat.bipartite()){
-                this->sgn_kappa_ = -1;
-            } else {
+            if(this->Lat.bipartite()){
                 this->sgn_kappa_ = +1;
+            } else {
+                this->sgn_kappa_ = -1;
             }

```

Comment:
@luutom do we actually need to put in the species-dependent sign on bipartite lattices?  IIRC it doesn't actually matter for fermion determinant positive-definiteness.

I'm nervous about it because I think if you put this sign here we've got to propagate it into observables too.

Reply:
Done with 1c920e3c005f01acb39be86fbc22252e577226d4